private/bufpkg/bufimage/bufimagemodify/internal/internal.go

```diff
@@ -83,8 +81,18 @@ func RemoveLocationsFromSourceCodeInfo(sourceCodeInfo *descriptorpb.SourceCodeIn
 	// are structured as expected, and collect all of the indices that
 	// we need to delete.
 	indices := make(map[int]struct{}, len(paths)*2)
+	// each path in this trie is for a FieldOptions message (not for a singular option)
+	fieldOptionsPaths := fieldOptionsTrie{}
 	for i, location := range sourceCodeInfo.Location {
-		if _, ok := paths[GetPathKey(location.Path)]; !ok {
+		path := location.Path
+		if isPathForFieldOptions(path) {
+			fieldOptionsPaths.insert(path, i)
+		}
+		if _, ok := paths[GetPathKey(path)]; !ok {
+			if isPathForSomeFieldOption(path) {
+				// This field option path is not marked, register it to its parent FieldOptions.
```

Comment:
I guess "not marked" is what is meant by `!ok` in the condition? Does that mean that `paths` contains the items being _removed_? (I believe we should only register them if they are _not_ being removed.)

If so... that's kinda confusing. Usually a mark+sweep algorithm marks the live set, and then you sweep away everything unmarked. But is this instead marking the things to remove? (Please forgive my ignorance of what this package is or how it works.)

If `paths` are the items being removed, maybe call it `removedPaths` for clarity? And, if not, maybe `markedPaths`?

Reply:
That was my mistake, the helper doesn't need to know the concept of marking (the caller is a `MarkSweeper`). I updated it to `pathsToRemove` for clarity.

For context, in `buf generate`, if managed mode is enabled, it will call some number of `ModifyXYZ` functions, each `XYZ` being a file option or field option (we have a defined set of what `XYZ` can be).

For example, the branch that this branch is merging into, `mm-jstype`, implements `ModifyJSType`, which [modifies the jstype option on the compiled descriptor](https://github.com/bufbuild/buf/pull/2311/files#diff-0d68cfe9b7b680476f1736209b375fe7876766a166419efbfd591632d2a700c4R97), and then lets a `MarkSweeper`  mark the path to this field option. Once all such options are modified on the compiled descriptor set, managed mode let `MarkSweeper` sweep based on these marked paths, removing these paths from source code info locations. That is what this PR is about, expanding the capability of the sweeper to remove the paths to `FieldOptions` when all children field options are removed.

Let me know if you'd like to know more!